Docs: Monolithic architecture and its challenges (#10)
FuelLabs/fuel-docs: src/monolithic.md

# Monolithic Architecture 

Blockchains as we know them have four functions. In no particular order:
- Execution: Execute transactions to make updates to the state.
- Settlement: Dispute resolution.
- Consensus: Defines the state and validates that all nodes on the blockchain have the same state.
- Data availability: Ensure data stored in each block is available.

Monolithic blockchains are a type of blockchain architecture that handle all four functions, at the same time, on this single layer. 

> ![monolithic](https://user-images.githubusercontent.com/15346823/187779430-050becc8-702c-4836-a206-c4151811ecdb.png)

Reviewer: Can this be turned into a local image?

Author: done

# Monolithic Architecture

Blockchains as we know them have four functions. In no particular order:

- Execution: Execute transactions to make updates to the state.
- Settlement: Dispute resolution.
- Consensus: Defines the state and validates that all nodes on the blockchain have the same state.

> - Data availability: Ensure data stored in each block is available.

Reviewer: ```suggestion
- Data availability: Ensure block data has been published to the network.
```

Monolithic blockchains are a type of blockchain architecture that handle all four functions, at the same time, on this single layer.

![monolithic](./images/monolithic.png)

## Challenges with Monolithic

Some constraints and challenges with a monolithic architecture:

### Costly and inefficient transaction verification

In order to verify the validity of transactions in the chain, full nodes must download the entire chain and execute each transaction locally.

### Not designed for fraud proofs

> Using the general construction for fraud proofs, in order to create a fraud proof you have to provide the pre-state and compute the post-state, comparing what the block producer's post-state was to their outcome. If a contract calls other contracts, the pre-state could be extremely large, and the cost to produce a fraud proof could be unbound.

Reviewer: This section, while true in isolation, isn't really a related to monolithic vs modular. I would instead have a section after the next section (or expand the next section) on how using up the node's capacity for execution means there's less capacity left over for data availability.

> ### Resource constrains

Reviewer: ```suggestion
### Resource constraints
```

> The blockchain is bound by the resource capacity of its nodes. Throughput is constrained by the resource requirements of a *single* node. Since the blockchain is replicated, not distributed, across nodes.

Reviewer: ```suggestion
The blockchain is bound by the resource capacity of its nodes. Throughput is constrained by the resource requirements of a _single_ node since the blockchain is replicated, not distributed, across nodes.
```

### Scalability

> Scalability is defined as throughput / decentralization. To increase throughput, the number transactions per second, you have to increase bandwidth capacity, which pushes up the cost to run a full node as a user. This is not scalability, as it reduces the number of people with capacity to run a full node to validate the chain.

Reviewer: ```suggestion
Scalability is defined as the ratio of throughput to decentralization. To increase throughput—the number transactions per second—you have to increase bandwidth, compute, and storage capacity, which pushes up the cost to run a full node as a user. This is not scalability, as it reduces the number of people who can run a full node to validate the chain.
```
